Handle recovery from late out-of-band error arrivals

jreineckearm opened the pull request:
Address #402 , but actually goes beyond to fix the root cause.

Problem: If you set too many breakpoints
* gdb may report a completed (e.g.) `continue` command as `running`/`done`.
* gdb/gdbserver may install breakpoints and start running.
* Along the way, gdbserver may reject adding breakpoints if not enough HW resource is available.
* gdb may send a late `error` response referring to the previously successfully completed `continue` command.

Before this change, this could heavily confuse the adapter states and provoke funny behavior.

This PR changes this to recover from such situations. It
* stores additional, artificial information with the MI result data in the MI parser, prefixed with `cdt-` to avoid confusion with real MI payload
  * 'cdt-token': token of a command
  * 'cdt-command': original command string of a completing command that is still in the queue.
* Changes the MI parser command queue to add this data.
* Adds a `resultAsync` event between MI parser and session classes sending GDB async results. 'cdt-command' is only added for the first result for a token.
* Extends the thread state to include the token of the last command setting it running. (Cleared on stopped notifications)
  * Handle `done`/`running` results of successful commands that resume the target (sorry about that list of potential commands, I know it's ugly). Store their token per thread.
  * Handle `error` results for commands with tokens stored in the thread states. These are the out-of-band errors and require returning internal and VS Code thread states to stopped.


Note: Still a draft as I need to add a few pinpointed tests.

cwalther commented:
No worries, you are not obligated to respond quickly. From a quick merge test there doesn’t seem to be any overlap with what I’m doing, and if there is, I will adapt.

jreineckearm commented on src/gdb/GDBDebugSessionBase.ts:
```diff
@@ -2047,6 +2092,87 @@ export abstract class GDBDebugSessionBase extends LoggingDebugSession {
+                    }
+                    updateThreads.forEach(
+                        (thread) =>
+                            (thread.lastRunToken = notifyData['cdt-token'])
```
Thanks! Well spotted, indeed doesn't make sense to access `notifyData` again.

thorstendb-ARM reviewed (approved):
Tested on Jens' branch with STLink on pyOCD.
- Before: gdb session crashes with a strange error message and is not reacting anymore (must be killed).
- After: "HW breakpojnt limit reached, ..." is reported and session remains reactive. After removing some BPs, session runs as expected. Supports up to 6 BPs (or 5 + tempBP @ main)

<img width="971" height="292" alt="image" src="https://github.com/user-attachments/assets/35fb16d2-31fd-4693-8b0c-29dbe8352066" />

jreineckearm commented:
Rebased

jreineckearm commented:
Preparing a 1.2.0 release without this change. This needs more testing for the `non-stop` mode. Or alternatively a guard to apply only to all-stop mode in first incarnation.

cwalther commented on src/integration-tests/lateAsyncErrorsRemote.spec.ts:
```diff
@@ -0,0 +1,158 @@
+        });
+
+        // Stopped by default
+        let stoppedEvent: DebugProtocol.StoppedEvent | undefined = undefined;
+        await Promise.all([
+            new Promise<void>(async (resolve) => {
+                stoppedEvent = (await dc.waitForEvent(
+                    'stopped'
+                )) as DebugProtocol.StoppedEvent;
+                resolve();
+            }),
+            dc.configurationDoneRequest(),
+        ]);
```
This looks overcomplicated, why not
```ts
        let [stoppedEvent] = await Promise.all([
            dc.waitForEvent('stopped'),
            dc.configurationDoneRequest(),
        ]);
```
?

(Just some bikeshedding from a quick glance, not a thorough review.)

jreineckearm replied:
Updated in https://github.com/eclipse-cdt-cloud/cdt-gdb-adapter/pull/415/commits/27de94ae3c2de734a31ff22f21bd02a2f2f452cf

cwalther replied:
Heh, so do I! We’re learning together. 🙂

jreineckearm commented on src/integration-tests/lateAsyncErrorsRemote.spec.ts:
```diff
@@ -0,0 +1,158 @@
+
+    beforeEach(async function () {
+        dc = await standardBeforeEach('debugTargetAdapter.js');
+        await dc.launchRequest(
```
Awesome, thanks for reviewing! Was about to ask if you could have a look as you are looking into these at the moment. 👍 
If it's not too much extra work for you later, I'd like to get this fix in today or early next week. Planning to make a release next week and finishing this up was pending way too long on my ToDo list.

jreineckearm commented:
Rebased and added a test for the command queue and state recovery over DAP.
No functional change, already received feedback on added test.
Waiting for re-review of my changes, planning to merge latest on Monday.
